File: src/web/routes/potRoutes.ts
/**
 * POT Provider endpoints (bgutil-ytdlp-pot-provider compatible).
 * These endpoints allow yt-dlp to use Moombox as a POT provider.
 */

import type { Application, Request, Response, NextFunction } from "express";
import { Logger } from "../../core/logger.js";
import { PotProvider } from "../../core/potProvider.js";
import { asyncHandler } from "./errorHandler.js";
import { createRateLimiter } from "./rateLimiter.js";
import { isLoopback } from "../../utils/ipValidation.js";

// POT Provider version for compatibility
const POT_PROVIDER_VERSION = "1.0.0";

/** Restrict access to loopback IPs only */
function requireLoopback(req: Request, res: Response, next: NextFunction): void {
  const ip = req.ip || req.socket.remoteAddress || "";
  if (isLoopback(ip)) {
    next();
  } else {
    res.status(403).json({ error: "Forbidden: loopback only" });
  }
}

const potRateLimiter = createRateLimiter(10, 60 * 1000); // 10 req/min

export function registerPotRoutes(app: Application): void {
  const logger = Logger.getInstance();

  // Generate PO Token - main endpoint for yt-dlp (loopback only, rate limited)
  app.post("/get_pot", requireLoopback, potRateLimiter, asyncHandler(async (req, res) => {
    const body = req.body || {};

    // Handle deprecated parameters (must reject before schema validation for clear error messages)
    if (body.data_sync_id) {
      logger.warn(
        "[PotProvider] Rejected request using deprecated data_sync_id",
      );
      return res.status(400).json({
        error: "data_sync_id is deprecated, use content_binding instead",
      });
    }
    if (body.visitor_data) {
      logger.warn(
        "[PotProvider] Rejected request using deprecated visitor_data",
      );
      return res.status(400).json({
        error: "visitor_data is deprecated, use content_binding instead",
      });
    }

    const contentBinding: string | undefined = body.content_binding;
    const bypassCache: boolean = body.bypass_cache === true;

    logger.info(
      `[PotProvider] POT requested${contentBinding ? ` for ${contentBinding.substring(0, 20)}...` : " (no binding)"}${bypassCache ? " (bypass cache)" : ""}`,
    );

    const potProvider = PotProvider.getInstance();
    const sessionData = await potProvider.generatePoToken(
      contentBinding,
      bypassCache,
    );

    logger.info(
      `[PotProvider] POT generated: ${sessionData.poToken.substring(0, 30)}...`,
    );
    res.json(sessionData);
  }));

  // Invalidate all caches (loopback only — state-changing)
  app.post("/invalidate_caches", requireLoopback, (req, res) => {
    logger.info("[PotProvider] Cache invalidation requested");
    const potProvider = PotProvider.getInstance();
    potProvider.invalidateCaches();
    res.status(204).send();
  });

  // Invalidate integrity tokens (loopback only — state-changing)
  app.post("/invalidate_it", requireLoopback, (req, res) => {
    logger.info("[PotProvider] Integrity token invalidation requested");
    const potProvider = PotProvider.getInstance();
    potProvider.invalidateIntegrityTokens();
    res.status(204).send();
  });

  // Ping endpoint for health checks
  app.get("/ping", (req, res) => {
    logger.debug("[PotProvider] Ping received");
    res.json({
      server_uptime: process.uptime(),
      version: POT_PROVIDER_VERSION,
    });
  });

  // Debug endpoint to view minter cache
  app.get("/minter_cache", (req, res) => {
    logger.debug("[PotProvider] Minter cache requested");
    const potProvider = PotProvider.getInstance();
    const cache = potProvider.getMinterCache();
    res.json(Array.from(cache.keys()));
  });
}

File: internal/web/routes/pot_routes.go
package routes

import (
	"context"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/vampiricwulf/Moombox/internal/web"
)

// PotSessionData represents a PO token response.
type PotSessionData struct {
	PoToken        string `json:"poToken"`
	ContentBinding string `json:"contentBinding"`
}

// PotProviderInterface defines the PO token provider methods needed by routes.
type PotProviderInterface interface {
	GeneratePoTokenString(ctx context.Context, contentBinding string, bypassCache bool) (string, error)
	GeneratePoTokenSession(ctx context.Context, contentBinding string, bypassCache bool) (poToken string, actualBinding string, err error)
	InvalidateCaches()
	InvalidateIntegrityTokens()
	GetMinterCacheKeys() []string
}

// PotRoutesDeps holds dependencies for POT routes.
type PotRoutesDeps struct {
	PotProvider PotProviderInterface
	StartTime   time.Time
	RateLimit   *web.RateLimiter
	Logger      interface {
		Debug(msg string, args ...any)
		Info(msg string, args ...any)
		Warn(msg string, args ...any)
	}
}

// PotRoutes registers POT provider endpoints (root-mounted).
func PotRoutes(r chi.Router, deps *PotRoutesDeps) {
	potRL := deps.RateLimit

	r.With(web.LoopbackOnly, potRL.Middleware).Post("/get_pot", func(rw http.ResponseWriter, req *http.Request) {
		var body struct {
			ContentBinding string `json:"content_binding"`
			BypassCache    bool   `json:"bypass_cache"`
			DataSyncID     string `json:"data_sync_id"`
			VisitorData    string `json:"visitor_data"`
		}
		if err := json.NewDecoder(req.Body).Decode(&body); err != nil {
			jsonError(rw, "invalid request body", http.StatusBadRequest)
			return
		}

		// Reject deprecated fields (match TS: separate checks with warn logs)
		if body.DataSyncID != "" {
			if deps.Logger != nil {
				deps.Logger.Warn("[PotProvider] data_sync_id is deprecated, use content_binding instead")
			}
			jsonError(rw, "data_sync_id is deprecated; use content_binding", http.StatusBadRequest)
			return
		}
		if body.VisitorData != "" {
			if deps.Logger != nil {
				deps.Logger.Warn("[PotProvider] visitor_data is deprecated, use content_binding instead")
			}
			jsonError(rw, "visitor_data is deprecated; use content_binding", http.StatusBadRequest)
			return
		}

		if deps.PotProvider == nil {
			jsonError(rw, "PO token provider not available", http.StatusServiceUnavailable)
			return
		}

		cbLabel := ""
		if body.ContentBinding != "" {
			cbLabel = " (binding=" + body.ContentBinding + ")"
		}
		if deps.Logger != nil {
			deps.Logger.Info("[PotProvider] POT requested" + cbLabel)
		}

		poToken, actualBinding, err := deps.PotProvider.GeneratePoTokenSession(req.Context(), body.ContentBinding, body.BypassCache)
		if err != nil {
			if deps.Logger != nil {
				deps.Logger.Warn("failed to generate PO token", "error", err)
			}
			jsonError(rw, "failed to generate PO token", http.StatusInternalServerError)
			return
		}

		if deps.Logger != nil {
			hash := sha256.Sum256([]byte(poToken))
			deps.Logger.Info("[PotProvider] POT generated", "hash", fmt.Sprintf("%x", hash[:4]), "len", len(poToken))
		}

		jsonResponse(rw, &PotSessionData{
			PoToken:        poToken,
			ContentBinding: actualBinding,
		})
	})

	r.With(web.LoopbackOnly).Post("/invalidate_caches", func(rw http.ResponseWriter, req *http.Request) {
		if deps.Logger != nil {
			deps.Logger.Info("[PotProvider] Cache invalidation requested")
		}
		if deps.PotProvider != nil {
			deps.PotProvider.InvalidateCaches()
		}
		rw.WriteHeader(http.StatusNoContent)
	})

	r.With(web.LoopbackOnly).Post("/invalidate_it", func(rw http.ResponseWriter, req *http.Request) {
		if deps.Logger != nil {
			deps.Logger.Info("[PotProvider] Integrity token invalidation requested")
		}
		if deps.PotProvider != nil {
			deps.PotProvider.InvalidateIntegrityTokens()
		}
		rw.WriteHeader(http.StatusNoContent)
	})

	r.Get("/ping", func(rw http.ResponseWriter, req *http.Request) {
		if deps.Logger != nil {
			deps.Logger.Debug("[PotProvider] Ping received")
		}
		jsonResponse(rw, map[string]any{
			"server_uptime": time.Since(deps.StartTime).Seconds(),
			"version":       "1.0.0",
		})
	})

	// Audit Q-23: GET /minter_cache reveals internal POT cache keys.
	// Wrap in LoopbackOnly for parity with the other POT endpoints — this
	// is consumed by the local yt-dlp plugin only.
	r.With(web.LoopbackOnly).Get("/minter_cache", func(rw http.ResponseWriter, req *http.Request) {
		if deps.Logger != nil {
			deps.Logger.Debug("[PotProvider] Minter cache requested")
		}
		if deps.PotProvider == nil {
			jsonResponse(rw, []string{})
			return
		}
		keys := deps.PotProvider.GetMinterCacheKeys()
		if keys == nil {
			keys = []string{}
		}
		jsonResponse(rw, keys)
	})
}
